keps/sig-testing/2420-reducing-kubernetes-build-maintenance/README.md

```diff
@@ -0,0 +1,671 @@
+
+- In Kubernetes' CI we have bazel [remote caching] enabled which theoretically reduces our resource consumption and improves build times. In practice this gain is reduced currently, due to enabling multiple runs for unit tests to eliminate flakiness. Current measurements show equivalent runtime for the two builds.
+  - `go build` has developed high quality caching we can leverage if we need, we have a prototype of this already
+  - Kubernetes has enabled repeated runs of "unit" tests to surface flakes, which causes both `go` and Bazel to not cache test results
+  - We've disabled caching for large, poorly cachable objects like Docker images anyhow
+
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+1. Continue maintaining both build systems
+
+This has major drawbacks:
+
+- This continues to eat developer time without significant return, for the most part releases, the bulk of CI (periodic tests), and contributor development use the make build(s).
+
+- Key components of our development setup do not easily port to bazel, so it will continue to be an "also":
+  - code generators: Kubernetes uses a lot of generated code, while Bazel *can* run code generators just fine without checking in the generated sources, we need to check them in for consumption by non-bazel users (i.e. external projects), and as a result we have never leveraged this. These generators are largely incompatible with bazel "philosophy":
+    - much of the k8s codegen uses fake go build tags, which would require an external process (i.e. gazelle extension) to turn into build files
+    - additionally, some codegen relies on certain dependencies, which the gazelle extension would need to figure out
+    - many of the code generators are optimized to load the entire Go tree, then generate lots of code at once, which is somewhat incompatible with Bazel's approach to working on a package-by-package level. Generating code package-by-package is much slower (due to having to reparse the tree each time).
+  - separate "hack/tools" go module for linters etc: not only do linters not work well (because any source change busts the cache anyhow), but rules_go does not do multiple go modules well. Multiple go modules allows us to isolate development dependencies (like linters) from release binary dependencies, easing dependency management
+
+2. Improve bazel integration and drop the make based build
+
+In addition to the points made in 1.) above as to why this is not particularly viable for some of our existing development patterns:
+- We would need to improve support for CGO or eliminate CGO in Kubernetes, both of which would be somewhat expensive to develop
```

Comment:
Any other technical issue apart from cgo/rules_go integration (and the mentioned below consumed-by-go-get aspect)? It seems like the technical issues left to fix in rules_go/cgo integration are not insurmountable.

If these were to be tracked in an issue somewhere (and referred to here) it would make technical discussion easier.

Reply:
See other portions of the doc. CGO comments are completely unsupported by rules_go and have no intention of being supported because they don't meet bazel's concepts of "hermetic" (though C/C++ under bazel in general doesn't anyhow).

We depend on hacking around this in brittle ways currently:
https://github.com/kubernetes/kubernetes/blob/81d8ccfa8e641f7295af6dbfe5bbb3c157a10b55/hack/update-bazel.sh#L108-L113

https://github.com/bazelbuild/bazel-gazelle/issues/203#issuecomment-504397957
https://github.com/bazelbuild/bazel-gazelle/blob/987c29003d1616c0b2499fa7f5ce0f3bc11bc9a3/language/go/fileinfo.go#L399

Reply:
Note that this is not a technical issue to surmount: rules_go has outright rejected support. https://github.com/bazelbuild/bazel-gazelle/pull/556#issuecomment-504397812